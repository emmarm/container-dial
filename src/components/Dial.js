import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';

const DialButton = styled('button')({
  background: 'rgba(255, 255, 255, 0.9)',
  border: 'none',
  boxShadow:
    'rgba(50, 50, 93, 0.1) 0px 7px 14px, rgba(0, 0, 0, 0.08) 0px 3px 6px',
  cursor: 'pointer',
  display: 'grid',
  gridTemplateColumns: '1fr 80px',
  height: '100%',
  padding: 0,
  textDecoration: 'none',
  width: '100%',
  ':hover,:focus': {
    background: 'white'
  },
  '::-moz-focus-inner': {
    border: 0
  }
});

const TitleArea = styled('div')({
  alignItems: 'center',
  color: 'rgba(0,0,0,0.7)',
  display: 'flex',
  fontSize: 18,
  fontWeight: 400,
  height: 80,
  justifyContent: 'center',
  overflow: 'hidden',
  padding: 10,
  textAlign: 'center',
  wordBreak: 'break-word'
});

const Favicon = styled('div')(
  {
    height: 80,
    width: 80
  },
  ({ favicon, theme, useFavicon }) => ({
    background: useFavicon && `center / contain no-repeat url('${favicon}')`,
    color: theme.dark,
    fontSize: 70,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center'
  })
);

class Dial extends Component {
  state = { useFavicon: true };

  async componentDidMount() {
    const { favicon } = this.props.dial;

    const image = new Image();
    image.src = favicon;
    await image.complete;
    image.naturalHeight === 0 && this.setState(() => ({ useFavicon: false }));
  }

  getTrimmedName = siteName => {
    const lastWordEnd = siteName.lastIndexOf(' ', 33);
    return lastWordEnd > -1 && lastWordEnd < 33
      ? siteName.slice(0, lastWordEnd) + '...'
      : siteName.slice(0, 11) + '...';
  };

  getLetter = siteName => siteName.charAt(0).toUpperCase();

  openLink = () => {
    window.location = this.props.dial.siteUrl;
  };

  render() {
    const {
      dial: { favicon, siteName }
    } = this.props;
    const { useFavicon } = this.state;
    return (
      <DialButton onClick={this.openLink} tabIndex={1}>
        <TitleArea>
          {siteName.length > 36 ? this.getTrimmedName(siteName) : siteName}
        </TitleArea>
        <Favicon favicon={favicon} useFavicon={useFavicon}>
          {!useFavicon && this.getLetter(siteName)}
        </Favicon>
      </DialButton>
    );
  }
}

Dial.propTypes = {
  dial: PropTypes.shape({
    container: PropTypes.string,
    favicon: PropTypes.string,
    id: PropTypes.number,
    siteName: PropTypes.string,
    siteUrl: PropTypes.string
  })
};

export default Dial;
