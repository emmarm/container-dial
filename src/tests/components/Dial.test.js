import React from 'react';
import { shallow } from 'enzyme';

import Dial from '../../components/Dial';

describe('Dial Component', () => {
  const props = { siteName: 'test', icon: 'icon.jpg' };
  const wrapper = shallow(<Dial {...props} />);

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('applies name prop correctly', () => {
    const title = wrapper.find('.dial__title');
    expect(title.text()).toBe(props.siteName);
  });

  it('applies icon prop correctly', () => {
    const icon = wrapper.find('.dial__icon');
    expect(icon.prop('style')).toEqual({
      backgroundImage: `url('${props.icon}')`
    });
  });
});
