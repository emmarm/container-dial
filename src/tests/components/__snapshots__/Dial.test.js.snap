// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dial Component matches snapshot 1`] = `
<Styled(button)
  onClick={[Function]}
  tabIndex={1}
>
  <Styled(div)>
    Goodreads
  </Styled(div)>
  <Styled(div)
    favicon="icon.jpg"
    useFavicon={false}
  >
    G
  </Styled(div)>
</Styled(button)>
`;
